### PR TITLE
Docs: Document handling eviction requests

### PR DESCRIPTION
**Pull request by orelmisan**
### What this PR does
Before this PR:
One needs to dig through a lot of code in order to understand how KubeVirt handles K8s' API-initiated eviction [1].

After this PR:
The way KubeVirt handles K8s' API-initiated eviction is documented.

[1] https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: docs/handling-eviction-requests.md
Comment (dankenigsberg) [inline docs/handling-eviction-requests.md:75]: Why do we need two mechanisms? Why pdb is not enough?

File: docs/handling-eviction-requests.md
Comment (dankenigsberg) [inline docs/handling-eviction-requests.md:78]: Why are we using two terms? Is there a difference between Evacuation and Eviction?

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
Thanks for enlightening me. Would you elaborate a bit more?

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:75]: The PDB just blocks the eviction request. I think it is still in-use as a precaution in case the webhook is unavailable, since the webhook is configured to be ignored in case of failure (failurePolicy: Ignore)[1].

The webhook marks the VMI for eviction.

[1] https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy

Comment (orelmisan) [reply docs/handling-eviction-requests.md:75]: The webhook cannot give the same as protection the PDB - which also protects two pods during migrations.

File: docs/handling-eviction-requests.md
Comment (ingvagabund) [inline docs/handling-eviction-requests.md:66]: How should be the table read? Either as
- If "Eviction strategy None" then ("Is VMI migratable" does not apply, "VM is not marked for eviction", "Webhook approves eviction", ...)

or

- If ("Eviction strategy" == None, "Is VMI migratable" can be of any value, "Is VMI marked for eviction" == false, "Does Webhook approve eviction" == true, "Does PDB allow eviction" == true) then Response

or differently?

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md]: The Eviction strategy and Whether the VMI is migratable are the inputs.
The `Is VMI marked for eviction`, `Does Webhook approve eviction` and `Response` columns are the result.
I've added an explanation, please tell me if it is more readable now.

Changed `N/A` to `True/False`.
N/A is not applicable.

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:57]: Yes.

I also added a note saying that the migration controller also expand this PDB to both source and target pods during migration.

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:74]: Depending on the eviction strategy and is VMI migratable - the webhook decides what to do.
Both the webhook and the PDB should approve the eviction in order to receive 200, otherwise 429 is returned.

**Comment by ingvagabund** [comment]
Looks good. Thank you for addressing the comments.

File: docs/handling-eviction-requests.md
Comment (dankenigsberg) [reply docs/handling-eviction-requests.md:83]: Would you explain why two different mechanisms are needed? What is the problems and benefits of each of them? Please include the answer in the text of this documentation.

Comment (orelmisan) [reply docs/handling-eviction-requests.md:83]: Please see the note under `Pod Eviction Webhook`, it should answer this question.

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:86]: The name `Evacuation controller` is AFAIU historic.
It is also present in [VMI.status.evacuationNodeName](https://kubevirt.io/api-reference/main/definitions.html#_v1_virtualmachineinstancestatus).

In KubeVirt CR [spec.configuration](https://kubevirt.io/api-reference/main/definitions.html#_v1_kubevirtconfiguration), we have `evictionStrategy`.

So it seems that unfortunately there are two names which refer to the same process.
IMO this naming issue deserves a dedicated discussion.
WDYT?

Comment (dankenigsberg) [reply docs/handling-eviction-requests.md:86]: I believe that this doc should provide the explanation you just gave.

Comment (orelmisan) [reply docs/handling-eviction-requests.md:86]: After another reading of the PR description of PR #2091, it seems that both terms are there from day one.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/7ab39f682a7a0b82fd3cc33e41201368adb46ee7..70cb759aca74f54e0e4965008ece09156897a7da): Answered @dankenigsberg comments.

File: docs/handling-eviction-requests.md
Comment (dankenigsberg) [reply docs/handling-eviction-requests.md:89]: Thanks.

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:26]: The descriptions in this table were copied from [here](https://github.com/kubevirt/kubevirt/blob/9c3a52d67c6782cd48ae0702df5493f5023a04cc/staging/src/kubevirt.io/api/core/v1/types.go#L112)

> PDB plays the same role as in other strategies, no?

Yes, it does

> What is the value that is set on vmi.Status.EvacuationNodeName ?

The node name on which the virt-launcher is currently scheduled on - and has to be evacuated from.

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:29]: IMO its something that is important to emphasize - this webhook sees all eviction requests, and there is currently no way of filtering just virt-launchers.

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:39]: AFAIK, there could be multiple webhooks (called in a random order).
Approving the request does not mean the end of the process - there could be additional webhooks and a PDB.

There is also a [KEP](https://github.com/kubernetes/enhancements/pull/4213) about this subject.

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:26]: I suggest to perform the change both in the code in this doc in a follow-up.
WDYT?

**Comment by orelmisan** [comment]
@dankenigsberg what do you think is missing in order to move forward with the PR?

**Comment by dankenigsberg** [comment]
>  dankenigsberg what do you think is missing in order to move forward with the PR?

I think it is mergeable, but I'm no expert about the subject matter - I cannot really tell if it is all true, and I was not sure I'm qualified to give it a

/lgtm

**Comment by orelmisan** [comment]
@xpivarc could you please take a look?

File: docs/handling-eviction-requests.md
Comment (xpivarc) [reply docs/handling-eviction-requests.md:39]: Maybe you want to rephrase to something like "the eviction request is passed back to chain"

**Review by xpivarc — COMMENTED** [review]
/approve
/hold

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/70cb759aca74f54e0e4965008ece09156897a7da..9c705b2983b17cdbb01d8806689bcf6f778c34d7): Answered @xpivarc's [comment](https://github.com/kubevirt/kubevirt/pull/11286#discussion_r1587559232).

File: docs/handling-eviction-requests.md
Comment (orelmisan) [reply docs/handling-eviction-requests.md:43]: Done.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/9c705b2983b17cdbb01d8806689bcf6f778c34d7..4a93ff5fbc6973bc5099517daf3739a669309492): Fixed typos
Thank you @enp0s3

**Review by enp0s3 — COMMENTED** [review]
/approve

**Review by enp0s3 — APPROVED** [review]

**Pull request merged**
